subsys/bluetooth/mesh/main.c

```diff
@@ -88,6 +88,26 @@ int bt_mesh_provision(const u8_t net_key[16], u16_t net_idx,
 	return 0;
 }
 
+int bt_mesh_provision_adv(const u8_t uuid[16], u16_t net_idx, u16_t addr,
+			  u8_t attention_duration)
+{
+	if (!atomic_test_bit(bt_mesh.flags, BT_MESH_VALID)) {
+		return -EINVAL;
+	}
+
+	if (bt_mesh_subnet_get(net_idx) == NULL) {
+		return -EINVAL;
+	}
```

Comment:
One of the things that has been stressed at many Mesh WG meetings I've attended is that the Provisioner is not required to be a node on the network. It's only if we also support configuration client that we need to be a node. The above two checks seem to require that we're both provisioned into a network and that we've been configured with the net_idx we intend to give to the to-be-provisioned node.

What you should be checking here (I think) is if the net_idx is available in the provisioner's database.

While we could just say that we don't support separating configuration client and provisioner roles, we should still have separation of the provisioner's database and the parts of it that have been configured for the local node. We could e.g. be adding app & net keys to other nodes that the local node (configuration client) hasn't configured for its own use as a node (beyond passing values as parameters in configuration client messages).

Reply:
I think it would be a very good idea to have that sort of database that could have its own API for adding keys and such. The nodes added should probably also be put in that database instead of in the `struct bt_mesh_net`.

> What you should be checking here (I think) is if the net_idx is available in the provisioner's database.

Agree, but there is no notion of a provisioner's database currently. Do you think that is required before merging this PR?

The checks in the API call is basically just a "fail fast" but I'm pretty sure they could be removed. The real problem is that `send_prov_data()` uses the local node as database for the `net_key`, `iv_index` and `flags`. I think this is where the separation should happen.

For know I think it would be okay to say that separation is not supported but I see no problem of creating the separation in the future.

Reply:
@tsvehagen as long as there's no major impact on the public API I think we can do the separation (provisioner database) in a later PR, but it'd be good to be done sooner than later, ideally before the 2.1 merge window closes in a couple of weeks.

Reply:
Okay so regarding the API I think we should then have a short discussion before merging.

Currently the API only takes a `net_idx`. This won't be enough if for example a provisioner can work with multiple networks as a the same `net_idx` could appear in both. In that case there must also be some sort of network id parameter. Is that something we want to support?

Reply:
> One of the things that has been stressed at many Mesh WG meetings I've attended is that the Provisioner is not required to be a node on the network. It's only if we also support configuration client that we need to be a node. The above two checks seem to require that we're both provisioned into a network and that we've been configured with the net_idx we intend to give to the to-be-provisioned node.

@jhedberg 
One thing that I don't really understand is that during provisioning, like I commented above, the provisioning data includes `flags` and `iv_index`. If the provisioner device is separate from the configuration client node, how will the provisioner know these things? Is there anything specified for how a provisioner can get new network information?

Reply:
>  If the provisioner device is separate from the configuration client node, how will the provisioner know these things? Is there anything specified for how a provisioner can get new network information?

This is left as a vendor-specific solution for the provisioner implementation. There's a spec coming out soon with a standard format for encoding such information, which can be used to synchronise multiple provisioners of the same network (this e.g. lets provisioners get subsets of the unicast address range so they don't go and provision different devices with the same addresses).

> Also in the Mesh Profile spec under 3.10.1 Mesh Network Creation procedure, it says that a Provisioner shall allocate a unicast address and assign it to its primary element. Doesn't this make it a node?

That's indeed quite odd, and in contradiction to what I've learned from listening and talking to  Mesh WG members. I'll check if upcoming spec versions have this clarified/corrected.

@trond-snekvik do you have any thoughts on these issues?

Reply:
> @trond-snekvik do you have any thoughts on these issues?

I'm also under the impression that the WG wants provisioners to be some omnipresent non-node entity, but I think the most natural way to implement it is to combine it with the configurator role.

I'm fine with the current API and its checks. Forcing the users to provide the netkey value in addition to the net_idx just introduces awkwardness for the normal case, and the current approach isn't really a blocker for the non-node provisioner.